contrib/mpi-proxy-split/split_process.cpp

```diff
@@ -328,6 +332,67 @@ startProxy()
   return childpid;
 }
 
+// Rounds the given address up to the nearest 2MB.
+static inline unsigned long long
+alignMemAddr(unsigned long long addr)
+{
+  return (addr + 0x1fffff) & ~0x1fffff;
+}
+
+// This sets the address range of the lower half by searching through the memory
+// region for the first available free region.
+static void
+dynamicMemRangeHelper(MemRange_t *lh_mem_range, const uint64_t *region_size)
+{
+  const uint64_t ONE_GB = 0x40000000;
+  
+  bool is_set = false;
+
+  char perms[8];
+  unsigned long offset;
+  char device[8];
+  unsigned long inode;
+  char prev_path_name[PATH_MAX];
+  char next_path_name[PATH_MAX];
+
+  unsigned long long prev_addr_start;
+  unsigned long long prev_addr_end;
+  unsigned long long next_addr_start;
+  unsigned long long next_addr_end;
+
+  // FIXME: fopen and fscanf may call malloc, which might call mmap.
+  // For now, we hope it doesn't change the calculation.
+  FILE* f = fopen("/proc/self/maps", "r");
+  if (f) {
+    fscanf(f, "%llx-%llx %s %lx %s %lx %s\n",
+           &prev_addr_start, &prev_addr_end, perms, &offset, device, &inode, prev_path_name);
+    while (fscanf(f, "%llx-%llx %s %lx %s %lx %s\n",
+           &next_addr_start, &next_addr_end, perms, &offset, device, &inode, next_path_name) != EOF) {
+      // alignMemAddr aligns the address such that HUGEPAGES/2MB can also be used.
+      prev_addr_end = alignMemAddr(prev_addr_end);
+
+      // We add a 1GB buffer between the end of the heap or the start of the stack.
+      if (strcmp(prev_path_name, "[heap]") == 0) {
+        prev_addr_end += ONE_GB;
+      }
+      if (strcmp(next_path_name, "[stack]") == 0) {
+        next_addr_start += ONE_GB;
+      }
+
+      if (prev_addr_end + *region_size <= next_addr_start) {
+        lh_mem_range->start = (VA)prev_addr_end;
+        lh_mem_range->end =   (VA)prev_addr_end + *region_size;
+        is_set = true;
+        break; 
+    }
+      prev_addr_end = next_addr_end;
+      strcpy(prev_path_name, next_path_name);
```

Comment:
Please don't use `strcpy`, one good practice I found is how git does it. https://github.com/git/git/blob/master/banned.h

Reply:
interesting. was your link supposed to have git's own implementation of `strcpy`? just says that it's banned.

since i'm changing the map-reading process to use `mtcp_readmapsline`, i'll also change this to use `mtcp_strcpy`.